Feature/iap over bt

KhrystynaDubovyk opened the pull request:
Provides the information for  #[1587](https://github.com/smartdevicelink/sdl_core/issues/1587)

Proposal: [Connectivity via iAP-BT and Transport Switch](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0053-Connectivity-via-iAP-BT-and-Transport-Switch.md)

This PR is **ready** for review.

### Summary
1. **TRS/iAP over BT/** folder was created
iAP_over_BT_data_resumption index was created  
iAP_over_BT_handling_HMI_notifications_during_re-connection index was created  
iAP_over_BT_re-connection index was created 
2. **detailed_docs/accessories/**  
Diagrams added
iAP transport change reconnection.png  
iAP_over_BT_data_resumption.png
3. **detailed_docs/iAP_over_BT/**  folder was created  
iAP_over_BT_handling_HMI_notif_during_reconnection_use_case index was created
iAP_over_BT_re-connection_use_case index was created

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

KhrystynaDubovyk committed:
TRS/iAP over BT/ folder was created
iAP_over_BT_data_resumption index was created
iAP_over_BT_handling_HMI_notifications_during_re-connection index was created
iAP_over_BT_re-connection index was created
detailed_docs/accessories/
Diagrams added
iAP transport change reconnection.png
iAP_over_BT_data_resumption.png
detailed_docs/iAP_over_BT/ folder was created
iAP_over_BT_handling_HMI_notif_during_reconnection_use_case index was created
iAP_over_BT_re-connection_use_case index was created

AStasiuk commented on detailed_docs/TRS/iAP over BT/iAP_over_BT_re-connection.md:
```diff
@@ -0,0 +1,77 @@
+send BC.OnAppUnregistered (appID, unexpectedDisconnect:true) to HMI for such application
+
+4.
+In case mobile application sends RegisterAppInterface during `<AppTransportChangeTimer> + <AppTransportChangeTimerAddition>*N` timer  
```
@KhrystynaDubovyk , maybe better to use an 'app' instead of 'applications' as made in all rest of files?
It's for whole iAP_over_BT_re-connection.md

KhrystynaDubovyk replied:
Rejected.
changed 'app' into 'application' throughout  the document in favour of existing TRS documents written with 'application' (embedded_navi, RC, iAP_over_BT_data_resumptionTRS, ).

AStasiuk commented on detailed_docs/TRS/iAP over BT/iAP_over_BT_re-connection.md:
```diff
@@ -0,0 +1,77 @@
+SDL must:
+- consider such app as re-registered app
+- update data in PolicyTable (`<device_identifier>` section)
+- send UpdataDeviceList to HMI
```
@KhrystynaDubovyk , type 'UpdataDeviceList', correct - 'UpdateDeviceList'.

KhrystynaDubovyk replied:
Accept. Done.

AStasiuk commented on detailed_docs/TRS/iAP over BT/iAP_over_BT_re-connection.md:
```diff
@@ -0,0 +1,77 @@
+_Note:_ In case of USB connection loss SDL must proceed with unexpected disconnect flow
+
+## Non-Functional Requirements
+New parameters "AppTransportChangeTimer" , "AppTransportChangeTimerAddition"   must be added to .ini file -> [MAIN] section
```
@KhrystynaDubovyk, suggested to indicate directly: smartdevicelink.ini file

AStasiuk commented on detailed_docs/TRS/iAP over BT/iAP_over_BT_handling_HMI_notifications_during_re-connection.md:
```diff
@@ -0,0 +1,44 @@
+discard all responses to such `appID` from HMI
+
+## Non-Functional Requirements
+New parameters "AppTransportChangeTimer" , "AppTransportChangeTimerAddition"   must be added to .ini file -> [MAIN] section
```
@KhrystynaDubovyk, suggested to indicate directly: smartdevicelink.ini file

KhrystynaDubovyk replied:
Accepted. Done.

AStasiuk commented on detailed_docs/TRS/iAP over BT/iAP_over_BT_data_resumption.md:
```diff
@@ -0,0 +1,55 @@
+behave according to existing Data Resumption rules 
+
+## Non-Functional Requirements
+New parameters "AppTransportChangeTimer" , "AppTransportChangeTimerAddition"   must be added to .ini file -> [MAIN] section
```
@KhrystynaDubovyk, suggested to indicate directly: smartdevicelink.ini file

KhrystynaDubovyk replied:
Accepted. Done.

KhrystynaDubovyk committed:
Update iAP_over_BT_data_resumption index

changing ".ini file" into "smartdevicelink.ini file"

KhrystynaDubovyk committed:
Update iAP_over_BT_handling_HMI_notifications_during_re-connection index

changed .ini file into smartdevicelink.ini file

KhrystynaDubovyk committed:
Update iAP_over_BT_re-connection index

fix typo
change .ini file into smartdevicelink.ini file

KhrystynaDubovyk committed:
Update iAP_over_BT_re-connection index

Minor text formatting;  smoothing  'app'/'application'  usage throughout  the document

KhrystynaDubovyk commented on detailed_docs/iAP_over_BT/iAP_over_BT_handling_HMI_notif_during_reconnection_use_case.md:
```diff
@@ -0,0 +1,26 @@
+
+_Pre-conditions:_  
+a. iOS device is connected over Bluetooth  
+b. N applications from this device are registered in SDL  
```
Rejected.
Changed 'app' into 'application' throughout  iAP_over_BT_handling_HMI_notif_during_reconnection_use_case document in favour of existing use cases: 
**RC** UCs  are written mostly with 'app'
**embedded_navi** UCs are written mostly with 'application'
To mitigate difference in word usage in TRSs and UCs I offer to use 'application'.

KhrystynaDubovyk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+7. SDL receives `<RegisterAppInterface>` request **before** reconnection timer expires
+8. App sends **valid** "hashID" with `RegisterAppInterface` request 
+9. SDL sends succesful response to the app 
+10. SDL updates data in Policytable 
```
Rejected.
I offer to write **"PolicyTable"** in favour of existing docs written in this way.

KhrystynaDubovyk committed:
Update iAP_over_BT_re-connection_use_case

Minor text formatting
app -> application
Policytable -> PolicyTable
fixing typos

KhrystynaDubovyk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+
+_Exception 1:_  
+7.1. **After** timer expires app sends **valid** "hashID" with `<RegisterAppInterface>` request  
+7.1.a. SDL notifies HMI about app being unregistred   
```
Accepted. Done.

KhrystynaDubovyk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+
+_Pre-conditions:_  
+a. iOS device is connected over USB  
+b. App from iOS device is registred and running on SDL  
```
Accepted. Done.

KhrystynaDubovyk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+_Expected:_   
+
+2. SDL does not start reconnection timer  
+3. SDL notifies HMI about app being unregistred
```
Accepted. Done.

KhrystynaDubovyk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+**Use Case 3**  
+_Pre-conditions:_  
+a. iOS device is connected over Bluetooth  
+b. App `<app_1>` from iOS device is registred and running on SDL  
```
Accepted. Done.

KhrystynaDubovyk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+ 1.1.b SDL receives signal with `<UUID_2>` about Bluetooth connection of another device  
+ 1.1.c SDL opens hub EAP protocol for `<UUID_1>` USB session  
+ 1.1.d SDL closes active EAP session opened for `<UUID_1>` Bluetooth  
+ 1.1.e SDL opens hub EAP protocol for `<UUID_2>` BT Bluetooth session  
```
Accepted. Done

KhrystynaDubovyk committed:
Update iAP_over_BT_re-connection_use_case index

edit headers fot UC2, 3
add titles for  UC2, 3

KhrystynaDubovyk commented:
@robinmk, please review

robinmk commented on detailed_docs/TRS/iAP over BT/iAP_over_BT_data_resumption.md:
````diff
@@ -0,0 +1,55 @@
+New parameters "AppTransportChangeTimer" , "AppTransportChangeTimerAddition"   must be added to smartdevicelink.ini file -> [MAIN] section
+
+```
+[MAIN] 
````
I think this can be rewritten as 
_Defines the timeout period SDL Core will wait for an application to re-register via alternate transport during 'transport switch'. If the application fails to re-register within this time, SDL Core shall inform HMI that the application has unregistered._

KhrystynaDubovyk replied:
Done.
"AppTransportChangeTimer" description was rewritten.

robinmk commented on detailed_docs/TRS/iAP over BT/iAP_over_BT_data_resumption.md:
```diff
@@ -0,0 +1,55 @@
+
+## Diagram
+Data resumption
+![Data resumption](https://github.com/smartdevicelink/sdl_requirements/blob/feature/iap_over_bt/detailed_docs/accessories/iAP_over_BT_data_resumption.png)
```
For the diagram, it would be very useful to indicate the event which causes SDL core to start the timer (AppTransportChangeTimer + AppTransportChangeTimerAddition*N)

KhrystynaDubovyk replied:
The diagram is updated

robinmk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+_Expected:_  
+
+2. SDL receives signal with `<UUID>` about USB connection of the same device 
+3. SDL compares `<UUID>` of BT device connection with `<UUID>` of USB connection  
```
Isn't the HMI responsible for doing the UUID check and informing SDL to perform transport switch? Please make sure to correct this entire flow accordingly

KhrystynaDubovyk replied:
The flow was corrected accordingly

robinmk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+7.1.c. SDL performs data resumption for this application  
+7.1.d. SDL **notifies** HMI about fresh application registration 
+
+_Exception 2:_  
```
What is the difference in case between Exception 2 and Exception 3?

KhrystynaDubovyk replied:
_Exception 2_ describes the flow when invalid or omitted hashID is received **after timer expires**;  
_Exception 3_ describes the flow when invalid or omitted hashID is received **before timer expires**

robinmk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+ 1. User connects the same device over USB
+ 2. SDL receives signal with `<UUID>` via IOSDeviceConnectInfo
+ 3. SDL connects USB device
+ 4. SDL disconnects Bluetooth
```
Step4-  What is meant by "SDL disconnects BT" - From my understanding it is the HMI layer which disconnects the iAP channel over BT transport (Please note that the BT transport itself is not closed, only the iAP channel over the BT transport is closed)

KhrystynaDubovyk replied:
Accepted, updated.

robinmk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+ 1.1. User connects another device over Bluetooth  
+ 1.1.a SDL receives signal with `<UUID_1>` about USB connection of the same device  
+ 1.1.b SDL receives signal with `<UUID_2>` about Bluetooth connection of another device  
+ 1.1.c SDL opens hub EAP protocol for `<UUID_1>` USB session  
```
1.1.c and 1.1.e- I think it is better to say that SDL opens appropriate EAP protocols and not be specific to HUB protocol because SDL should also be opening the dedicated EAP protocols as well (like com.smartdevicelink.multisession)

KhrystynaDubovyk replied:
Done.

robinmk commented on detailed_docs/iAP_over_BT/iAP_over_BT_re-connection_use_case.md:
```diff
@@ -0,0 +1,88 @@
+ 1.1.d SDL closes active EAP session opened for `<UUID_1>` Bluetooth  
+ 1.1.e SDL opens hub EAP protocol for `<UUID_2>` Bluetooth session  
+ 1.1.f SDL starts reconnection timer for `<UUID_1>`  
+ 1.1.g `<app_1>` from device `<UUID_2>` sends RegisterAppInterface (params)  
```
1.1.g - I am not sure I understand the case being described here. It seems like an app is requesting registration from the device which is connected over BT. But it is not clear why its registration request is being rejected. Please re-write this case with more clarity.

robinmk commented:
@KhrystynaDubovyk , I've provided the review comments.
Is this for requirements documentation?

KhrystynaDubovyk committed:
Updated iAP_over_BT_data_resumption

Changed
**Non-Functional Requirements**  
 `AppTransportChangeTimer` description.

KhrystynaDubovyk committed:
change description 'AppTransportChangeTimer'

KhrystynaDubovyk committed:
change 'AppTransportChangeTimer' description

KhrystynaDubovyk committed:
Update iAP_over_BT_re-connection_use_case

Update according to comment 
## Use Case 3: Multiple connection
_Exception 1:_
1.1.c
1.1.e

KhrystynaDubovyk committed:
Delete iAP_over_BT_data_resumption.png

KhrystynaDubovyk committed:
Upload updated iAP_over_BT_data_resumption.png

KhrystynaDubovyk committed:
Update iAP_over_BT_re-connection_use_case

Update Use Case 1, Use Case 3 according to comments.

KhrystynaDubovyk committed:
Update iAP_over_BT_re-connection_use_case.md
